src/features/elementfeature.jl

```diff
@@ -17,15 +37,16 @@ Construct a feature object that encodes features associated with individual atom
 struct ElementFeatureDescriptor <: AbstractAtomFeatureDescriptor
     name::String
     length::Integer
-    nbins::Integer
-    logspaced::Bool
+    encoder_decoder::EncoderDecoder
     categorical::Bool
     lookup_table::DataFrame
 end
 
 function ElementFeatureDescriptor(
     feature_name::String,
     lookup_table::DataFrame = atom_data_df;
+    encode_f::Function = default_efd_encode,
+    decode_f::Function = default_efd_decode,
```

Comment:
Even if they're populated by default, I'd rather not even have these be arguments in the constructor. Besides, they're basically fully specified by the fact that the `nbins`, `logspaced` kwargs are there anyway. If we were using some other hypothetical different encode fcn, we'd need different kwargs. If we really want it I suppose we could have an argument that's just a type of ED or something...

Reply:
Oh right, makes sense.

> If we really want it I suppose we could have an argument that's just a type of ED or something...

Yeah, I think that might be necessary since we need to have the `@assert`, but would probably be easier as a separate constructor altogether (since if you're specifying the ED you might not need `nbins`, and `logspaced`).

Reply:
> I'd rather not even have these be arguments in the constructor

Resolved in d88c264.

> Besides, they're basically fully specified by the fact that the nbins, logspaced kwargs are there anyway. If we were using some other hypothetical different encode fcn, we'd need different kwargs. If we really want it I suppose we could have an argument that's just a type of ED or something...

Yet to be resolved.

Reply:
> How would the length be defined in the case where data is not categorical, and an FD is constructed by passing an ED as argument (i.e., nbins aren't specified). Having length as a part of ED doesn't logically make sense, and hard-coding a check if the ED has a field named "nbins" feels like a very bad idea.

I'm realizing that the way things are used now, `nbins` and `length` should actually just be interchangeable I think. At one point, the actual `bins` were getting passed around more and there was a potential ambiguity between categorical/continuous (since in the former case, `length=length(bins)` while in the latter `length=length(bins)-1`) but now I think they're kind of the same, and maybe just confusingly named because `nbins` is not always equal to `length(bins)`. However, since as things stand now, `bins` is only ever generated/passed around internally and the user doesn't have to directly "handle" it, I think this is really just a naming issue rather than an actual procedural one, unless I'm missing something?

Maybe one way to make all this more elegant and keep functionalities separated would be not to have an actual field of the FD's that specifies the length but rather a function, something like:

```
vector_length(fd::AbstractFeatureDescriptor) = vector_length(fd.encoderdecoder)
vector_length(ed::OneHotED) = ed.categorical ? length(ed.bins) : length(ed.bins)-1
vector_length(ed::DirectED) = 1
```

I don't think ED's are currently storing the bin values directly, but that might not be a bad idea so that they don't have to be generated again every time we want to encode/decode. It also would help with the ambiguity about `nbins` and `length` that I mentioned above.

NB also, I called this `vector_length` to avoid confusion with `Base.length` which I don't think is really appropriate to dispatch here, but the most generic thing would probably be something more like `output_shape` to account for things like matrices that we'll use for OFM's etc. later.

Reply:
> vector_length(fd::AbstractFeatureDescriptor) = vector_length(fd.encoderdecoder)
vector_length(ed::OneHotED) = ed.categorical ? length(ed.bins) : length(ed.bins)-1
vector_length(ed::DirectED) = 1

Yeah, this looks better to me since it makes more sense for something like `vector_length`/`output_shape` to be more closely related to `EncoderDecoder` than `FeatureDescriptor`

> NB also, I called this vector_length to avoid confusion with Base.length which I don't think is really appropriate to dispatch here, but the most generic thing would probably be something more like output_shape to account for things like matrices that we'll use for OFM's etc. later

Yeah. We probably need to decide on a more "universal" and appropriate name too. #ntih

Reply:
oh actually, another thing that's definitely a property of a feature is the range of values it can take. So we could add a field for that, which for categorical things would just be a list (or maybe tuple?) and for continuous ones we could use Intervals.jl to specify a range?

Reply:
> Though I feel like the ED needs to know about the "categoricalness" also since the actual encoding process is different between categorical + continuous. So either that needs to be a field, or we split OneHotOneCold into two types, one for categorical and one for continuous...I don't have a strong feeling one way or the other, do you?

I don't think it needs to, since we can just pull it from the `efd` parameter passed to function defined in `encode_f`.
For instance, if `default_efd_encode` were the function in `encode_f`, then this distinction of categorical-ness could be made in the function itself, by pulling the value of `efd.categorical`.
```julia
function default_efd_encode(efd::ElementFeatureDescriptor, a::AbstractAtoms, nbins::Integer, logspaced::Bool)
```

Reply:
ahh true okay, I was forgetting that.

Reply:
> oh actually, another thing that's definitely a property of a feature is the range of values it can take. So we could add a field for that, which for categorical things would just be a list (or maybe tuple?) and for continuous ones we could use Intervals.jl to specify a range?

When you say "property of a feature", are you referring to `ElementFeature`s in particular?
I know OFM isn't really sketched out yet, but I can't really think of what "Intervals" might be valid for a FD like that in a tuple/vector...

Also, do you think this should be a field, or just a method defined on an `AbstractFeatureDescriptor`? I can't really think of any utility of having this defined as a field

Reply:
Well I think the notion of a set of valid values certainly extends to OFM (entries are always integers, bounded from below by 0 (inclusive) and upper bounds from sizes of shells). Defining it as a function could be tricky without getting back into dictionary hell since valid values for things like element or species features depend on what that feature is.

As I see it, the utility is twofold:
1) this information might be useful for a user to inspect to understand more about a particular feature
2) for direct encoding, the information will be needed to determine things like normalization or offsets (though I guess then it would be needed more directly by the ED than the FD, but as you pointed out in the other case, that info could be passed when the ED is constructed)

Reply:
> Defining it as a function could be tricky without getting back into dictionary hell since valid values for things like element or species features depend on what that feature is.

If the functions were defined something like -
```julia
function range(afd::AbstractFeatureDescriptor)
function range(efd::ElementFeatureDescriptor)
```
wouldn't the whole dictionary hell be avoidable? Here, we essentially have access to all the primary characteristics of the feature itself, using which we can arrive at the actual ranges themselves.
My reasoning for this is fairly straightforward. If `range` were a field in an `AbstractFD` like `ElementFD`, then it would probably be getting derived in the constructor itself. Here, we'd basically be doing the same steps, except we do it in `range()`, and save space by not making it a field.

> As I see it, the utility is twofold:

Having it as a function, would also satisfy these use-cases as well, right?